### ISSUE TITLE
DataLoader worker (pid xxx) is killed by signal

### ISSUE THREAD

**Issue by 1757525671** [open-mmlab/mmdetection#2406]
This error is reported when I use multithread training with multiple gpus, The parameters I specified are as follows,  img_pre_gpu=128, worker_pre_gpu=8, 2 gpu, This is a memory problem?

**Comment by 1757525671** [comment]
I checked the memory of the machine, and this program was killed after it took up 60g of memory. Why does it take up so much memory?

**Comment by hellock** [comment]
`worker_pre_gpu` is too large for you. There are 8 * 2 = 16 data workers, thus the memory is 16x.

**Comment by mdv3101** [comment]
Hi @hellock ,
I am getting the same error `RuntimeError: DataLoader worker (pid 14647) is killed by signal: Killed`.
The parameters I am using are: samples_per_gpu=2 and workers_per_gpu=1. I am training on two RTX2080Ti using dist_train.sh. Can you please suggest what is causing this error?

**Comment by mdv3101** [comment]
Hi @hellock 
I am again getting the error. Earlier It was working fine, but again it is showing this error. Can you please suggest what is causing this issue, so that atleast i can try to debug it.

```
Traceback (most recent call last):
  File "./tools/train.py", line 159, in <module>
    main()
  File "./tools/train.py", line 155, in main
    meta=meta)
  File "/home/ajoy.mondal/madhav/mmdetection/mmdet/apis/train.py", line 179, in train_detector
    runner.run(data_loaders, cfg.workflow, cfg.total_epochs)
  File "/home/ajoy.mondal/madhav/torch_env/lib/python3.7/site-packages/mmcv/runner/runner.py", line 383, in run
    epoch_runner(data_loaders[i], **kwargs)
  File "/home/ajoy.mondal/madhav/torch_env/lib/python3.7/site-packages/mmcv/runner/runner.py", line 282, in train
    self.model, data_batch, train_mode=True, **kwargs)
  File "/home/ajoy.mondal/madhav/mmdetection/mmdet/apis/train.py", line 74, in batch_processor
    losses = model(**data)
  File "/home/ajoy.mondal/madhav/torch_env/lib/python3.7/site-packages/torch/nn/modules/module.py", line 532, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/ajoy.mondal/madhav/torch_env/lib/python3.7/site-packages/torch/nn/parallel/distributed.py", line 447, in forward
    output = self.module(*inputs[0], **kwargs[0])
  File "/home/ajoy.mondal/madhav/torch_env/lib/python3.7/site-packages/torch/nn/modules/module.py", line 532, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/ajoy.mondal/madhav/mmdetection/mmdet/core/fp16/decorators.py", line 51, in new_func
    return old_func(*args, **kwargs)
  File "/home/ajoy.mondal/madhav/mmdetection/mmdet/models/detectors/base.py", line 148, in forward
    return self.forward_train(img, img_metas, **kwargs)
  File "/home/ajoy.mondal/madhav/mmdetection/mmdet/models/detectors/two_stage.py", line 145, in forward_train
    *rpn_loss_inputs, gt_bboxes_ignore=gt_bboxes_ignore)
  File "/home/ajoy.mondal/madhav/mmdetection/mmdet/models/dense_heads/rpn_head.py", line 49, in loss
    gt_bboxes_ignore=gt_bboxes_ignore)
  File "/home/ajoy.mondal/madhav/mmdetection/mmdet/core/fp16/decorators.py", line 131, in new_func
    return old_func(*args, **kwargs)
  File "/home/ajoy.mondal/madhav/mmdetection/mmdet/models/dense_heads/anchor_head.py", line 378, in loss
    featmap_sizes, img_metas, device=device)
  File "/home/ajoy.mondal/madhav/mmdetection/mmdet/models/dense_heads/anchor_head.py", line 135, in get_anchors
    featmap_sizes, device)
  File "/home/ajoy.mondal/madhav/mmdetection/mmdet/core/anchor/anchor_generator.py", line 193, in grid_anchors
    self.base_anchors[i].to(device),
  File "/home/ajoy.mondal/madhav/torch_env/lib/python3.7/site-packages/torch/utils/data/_utils/signal_handling.py", line 66, in handler
    _error_if_any_worker_fails()
RuntimeError: DataLoader worker (pid 38343) is killed by signal: Killed.
```

**Comment by mdv3101** [comment]
I was able to solve it by reducing `workers_per_gpu=0`

**Comment by tangsanli5201** [comment]
@hellock
I meet the same problem.  I used 2 V100 32GB  GPUS and tried 1/2/4 workers per_gpu, it goes well until training around 4k iterations. The same error occured: RuntimeError: DataLoader worker (pid xxxxx) is killed by signal: Killed. Reducing workers_per_gpu=0 (adviced by @mdv3101) is apparently not a enssential solution which decreases the training efficiency a lot.   
Will mmdet lead to CPU/GPU memory or multiprocessing leakage ? It is hard to explain why the dataloader will be killed after training some iterations.
 BTW, I found that the GPU-Util will increase during training (by cmd: nvidia-smi). It seemed that the when the GPU-Util reaches 100% usage, the error occured.
Looking forward to your response.
